src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -142,114 +141,179 @@ initialScannerState = ScannerState
-      scan <- use ssIgnore >>= \case
+      -> [ScannerM C.Node]
+      -> ScannerM C.Node
+    process pos ty subs = do
+      let node = C.Node pos ty []
+      use ssIgnore >>= \ign -> do
         -- When no `Ignore` state is set check next node for annotation,
         -- if found then set it as new `IgnoreMode` otherwise skip node.
-        Nothing                    -> handleIgnoreMode pos ty subs $ getIgnoreMode node
-        Just (Ignore mode modePos) ->
-          case (mode, ty) of
-            -- We expect to find a paragraph immediately after the
-            -- `ignore paragraph` annotanion. If the paragraph is not
-            -- found we should report an error.
-            (IMSParagraph, PARAGRAPH)    -> (ssIgnore .= Nothing) $> defNode
-            (IMSParagraph, x)            -> do
-              lift . tell . makeError modePos fp . ParagraphErr $ prettyType x
-              ssIgnore .= Nothing
-              Node pos ty <$> sequence subs
-
-            -- We don't expect to find an `ignore all` annotation here,
-            -- since that annotation should be at the top of the file and
-            -- the file should already be ignored when `checkIgnoreFile` is called.
-            -- We should report an error if we find it anyway.
-            (IMSAll, _)                 -> do
-              lift . tell $ makeError modePos fp FileErr
-              ssIgnore .= Nothing
-              Node pos ty <$> sequence subs
-
-            (IMSLink _, LINK {})         -> do
-              ssIgnore .= Nothing
-              return defNode
-            (IMSLink _, IMAGE {})        -> do
-              ssIgnore .= Nothing
-              return defNode
-            (IMSLink ignoreLinkState, _) -> do
-              when (ignoreLinkState == ExpectingLinkInSubnodes) $
-                ssIgnore . _Just . ignoreMode .=  IMSLink ParentExpectsLink
-              node' <- Node pos ty <$> sequence subs
-              when (ignoreLinkState == ExpectingLinkInSubnodes) $ do
-                currentIgnore <- use ssIgnore
-                case currentIgnore of
-                  Just (Ignore {_ignoreMode = IMSLink ParentExpectsLink}) -> do
-                    lift $ tell $ makeError modePos fp LinkErr
-                    ssIgnore .= Nothing
-                  _ -> pass
-              return node'
-
-      when (ty == PARAGRAPH) $ use ssIgnore >>= \case
+        let mbAnnotation = getAnnotation node
+        case mbAnnotation of
+          Just ann -> handleAnnotation pos ty ann
+          Nothing -> do
+            case ty of
+              PARAGRAPH -> handleParagraph ign pos ty subs
+              LINK {}   -> handleLink      ign pos ty subs
+              IMAGE {}  -> handleLink      ign pos ty subs
+              _         -> handleOther     ign pos ty subs
+
+    handleLink ::
+      Maybe Ignore ->
+      Maybe PosInfo ->
+      NodeType ->
+      [ScannerM C.Node] ->
+      ScannerM C.Node
+    handleLink ign pos ty subs = do
```

Comment:
Now all the `ign` arguments can renamed to something like `ann`. Right?

Reply:
Sorry, they always correspond to `Ignore`.

`<!-- ignore my-previous-comment -->`